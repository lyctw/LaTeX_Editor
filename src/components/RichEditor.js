import React, { Component } from 'react'
import LaTexEditor from './LaTexEditor'
import ReactQuill from 'react-quill'
import CustomToolbar from './CustomToolbar'
import '@instructure/canvas-theme'
import { Modal } from '@instructure/ui-overlays/lib/Modal'
import { Button, CloseButton } from '@instructure/ui-buttons'
import { Heading } from '@instructure/ui-elements/lib/Heading'
import 'react-quill/dist/quill.snow.css'


class RichEditor extends Component {
  constructor(props) {
    super(props)
    this.state = {
      content: '',
      open: false // modal open 
    }
  }

  // 1.BEGIN
  handleButtonClick = () => {
    this.setState(function (state) {
      return { open: !state.open }
    })
  };

  handleFormSubmit = e => {
    e.preventDefault()
    console.log('form submitted')
    this.setState(state => ({ open: false }))
  }

  renderCloseButton () {
    return (
      <CloseButton
        placement="end"
        offset="medium"
        variant="icon"
        onClick={this.handleButtonClick}
      >
        Close
      </CloseButton>
    )
  }

  renderMathModel() {
    return (
      <div style={{ padding: '0 0 11rem 0', margin: '0 auto' }}>
        {/* <Button onClick={this.handleButtonClick}>
          {this.state.open ? 'Close' : 'Open'} the Modal
        </Button> */}
        <Modal
          as="form"
          open={this.state.open}
          onDismiss={() => { this.setState({ open: false }) }}
          onSubmit={this.handleFormSubmit}
          size="auto"
          label="Modal Dialog: Hello World"
          shouldCloseOnDocumentClick
        >
          <Modal.Header>
            {this.renderCloseButton()}
            <Heading>Hello World</Heading>
          </Modal.Header>
          <Modal.Body>
              <LaTexEditor 
                handleInsert={(latex) => {
                  const imgTag = `<img src="https://latex.codecogs.com/svg.latex?${latex}">`;
                  this.setState((prevState) => {
                    return {
                      open: false, 
                      content: `${prevState.content} ${imgTag}`
                    }
                  })
                }}
              />
          </Modal.Body>
          {/* <Modal.Footer>
            <Button onClick={this.handleButtonClick}>Close</Button>&nbsp;
            <Button variant="primary" type="submit">Submit</Button>
          </Modal.Footer> */}
        </Modal>
      </div>
    )
  }
  // 1.END

  render() {

    const customModules = {
      toolbar: {
        container: '#toolbar'
      }
    }

    return (
      <div>
        <CustomToolbar 
          onToggleModal={this.handleButtonClick}
        />
        <ReactQuill
          id="quill"
          theme="snow"
          modules={customModules}
          // ref={this.props.editorRef}
          bounds="#right-nav"
          placeholder={this.props.placeholder}
          preserveWhitespace
          onChange={(content, delta, source, editor) => {
            this.setState({content})
            console.log(this.state)
            // console.log(editor.getHTML())
          }}
          value={this.state.content}
        />

        {this.renderMathModel()}
        
      </div>
    )
  }
}

export default RichEditor
