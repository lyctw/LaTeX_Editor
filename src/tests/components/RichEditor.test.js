import React from 'react'
import { shallow, mount, render } from 'enzyme'
import RichEditor from '../../components/RichEditor'
import toJson from 'enzyme-to-json';
import { JestEnvironment } from '@jest/environment';
 
test('should render RichEditor', () => {
    const wrapper = shallow(<RichEditor />);
    expect(toJson(wrapper)).toMatchSnapshot();
});

// test('should write Quill on input change', () => {
//     const text = '<p>something</p>';
//     const wrapper = shallow(<RichEditor />);
//     wrapper.find('Quill').simulate('change', {
//         content: text
//     });
//     expect(wrapper.state('content').toBe(text));
// });

test('should state.open is false by default', () => {
    const wrapper = shallow(<RichEditor />);
    expect(wrapper.state('open')).toBeFalsy();
});

test('should have a math button', () => {
    const wrapper = render(<RichEditor />);
    expect(wrapper.find('button.math').length === 1).toBe(true)
});


// // mount quill problem: see https://github.com/zenoamaro/react-quill/issues/434
// //                          https://stackoverflow.com/questions/48809753/testing-mutationobserver-with-jest
// test('should state.open is true if math button is clicked', () => {
//     const wrapper = mount(<RichEditor placeholder={'請在此輸入您的回覆'} getDocument={(html) => {jest.fn()}}/>);
//     wrapper.find('button.math').simulate('click', );
//     expect(wrapper.state('open')).toBeTruthy();
// });