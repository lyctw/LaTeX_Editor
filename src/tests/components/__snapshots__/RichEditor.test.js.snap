// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RichEditor 1`] = `
<div>
  <CustomToolbar
    onToggleModal={[Function]}
  />
  <Quill
    bounds="#right-nav"
    id="quill"
    modules={
      Object {
        "toolbar": Object {
          "container": "#toolbar",
        },
      }
    }
    onChange={[Function]}
    preserveWhitespace={true}
    theme="snow"
    value=""
  />
  <div
    style={
      Object {
        "margin": "0 auto",
        "padding": "0 0 0 0",
      }
    }
  >
    <Modal
      as="form"
      constrain="window"
      contentRef={[Function]}
      defaultFocusElement={null}
      insertAt="bottom"
      label="Modal Dialog: Hello World"
      liveRegion={null}
      mountNode={null}
      onClose={[Function]}
      onDismiss={[Function]}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      onOpen={[Function]}
      open={false}
      overflow="scroll"
      shouldCloseOnDocumentClick={true}
      shouldReturnFocus={true}
      size="large"
      transition="fade"
      variant="default"
    >
      <ModalHeader
        padding="small"
        variant="default"
      >
        <CloseButton
          buttonRef={[Function]}
          margin="0"
          offset="medium"
          onClick={[Function]}
          placement="end"
          size="small"
          variant="icon"
        >
          Close
        </CloseButton>
        <Heading
          border="none"
          color="inherit"
          ellipsis={false}
          level="h2"
        >
          Math Editor
        </Heading>
      </ModalHeader>
      <ModalBody
        as="div"
        padding="small"
        variant="default"
      >
        <Tabs
          elementRef={[Function]}
          margin="auto"
          onRequestTabChange={[Function]}
          padding="medium"
          shouldFocusOnRender={false}
          variant="default"
        >
          <Panel
            elementRef={[Function]}
            id="tabA"
            isDisabled={false}
            isSelected={true}
            labelledBy={null}
            padding="small"
            renderTitle="Edit Mode"
            textAlign="start"
            variant="default"
          >
            <LaTexEditor
              disableLatexEditor={[Function]}
              enableLatexEditor={[Function]}
              handleInsert={[Function]}
            />
          </Panel>
          <Panel
            elementRef={[Function]}
            id="tabB"
            isDisabled={false}
            isSelected={false}
            labelledBy={null}
            padding="small"
            renderTitle="Preview Mode"
            textAlign="start"
            variant="default"
          >
            <LaTexPreviewer
              disableLatexPreviewer={[Function]}
              enableLatexPreviewer={[Function]}
              handleInsert={[Function]}
            />
          </Panel>
        </Tabs>
      </ModalBody>
    </Modal>
  </div>
</div>
`;
